### PR TITLE
add vertex data ctx

### PR DESCRIPTION
**Pull request by pwrliang**

### CODE REVIEW

File: grape/app/vertex_data_context.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef GRAPE_APP_VERTEX_DATA_CONTEXT_H_
+#define GRAPE_APP_VERTEX_DATA_CONTEXT_H_
+#include "grape/grape.h"
```
Comment (yecol) [inline grape/app/vertex_data_context.h]: grape/grape.h is a header for examples,
include headers what you used.

File: grape/utils/default_allocator.h [hunk 2]
```diff
@@ -43,11 +43,11 @@ class DefaultAllocator {
   DefaultAllocator& operator=(DefaultAllocator&&) noexcept { return *this; }
 
   pointer allocate(size_type __n) {
-#ifdef __APPLE__
+// #ifdef __APPLE__
     return static_cast<pointer>(malloc(__n * sizeof(_Tp)));
-#else
-    return static_cast<pointer>(aligned_alloc(64, __n * sizeof(_Tp)));
-#endif
+// #else
```
Comment (yecol) [inline grape/utils/default_allocator.h]: ? revert this?

**Review by yecol — CHANGES_REQUESTED** [review]

File: grape/utils/default_allocator.h (hunk 2 above)
Comment (pwrliang) [reply grape/utils/default_allocator.h]: FIXME added

File: grape/app/vertex_data_context.h (hunk 1 above)
Comment (pwrliang) [reply grape/app/vertex_data_context.h]: fixed

**Comment by pwrliang** [comment]
Performance downgrade noticed on sparse dataset: https://file.alibaba-inc.com/lib/c9cf5c48-6d38-4958-a353-918bd009af1a/file/libgrape-lite.xlsx

**Comment by pwrliang** [comment]
The performance downgrade caused by initiation of VertexArray. Now, I have moved set_fragment to Worker::Init.

**Review by yecol — APPROVED** [review]

**Comment by pwrliang** [comment]
Benchmark for latest changes: https://file.alibaba-inc.com/lib/c9cf5c48-6d38-4958-a353-918bd009af1a/file/libgrape-lite.xlsx

File: examples/gnn_sampler/sampler_context.h [hunk 3]
```diff
@@ -29,14 +29,21 @@ limitations under the License.
 
 namespace grape {
 template <typename FRAG_T>
-class SamplerContext : public grape::ContextBase<FRAG_T> {
+class SamplerContext
+    : public VertexDataContext<FRAG_T, std::vector<typename FRAG_T::oid_t>> {
```
Comment (yecol) [inline examples/gnn_sampler/sampler_context.h:33]: For Sampler, the context should be TensorContext.

File: grape/app/vertex_data_context.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef GRAPE_APP_VERTEX_DATA_CONTEXT_H_
+#define GRAPE_APP_VERTEX_DATA_CONTEXT_H_
+#include "grape/app/context_base.h"
```
Comment (yecol) [inline grape/app/vertex_data_context.h:19]: insert blank lines in line-17,20

File: grape/app/vertex_data_context.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+#include "grape/utils/vertex_array.h"
+#define CONTEXT_TYPE_VERTEX_DATA "vertex_data"
```
Comment (yecol) [inline grape/app/vertex_data_context.h:22]: a more elegant way ?

**Review by yecol — CHANGES_REQUESTED** [review]

File: examples/gnn_sampler/sampler_context.h (hunk 3 above)
Comment (pwrliang) [reply examples/gnn_sampler/sampler_context.h:33]: TODO: https://github.com/yecol/libgrape-lite/issues/7

File: grape/app/vertex_data_context.h (hunk 5 above)
Comment (pwrliang) [reply grape/app/vertex_data_context.h:22]: We have to get context type by a base pointer

**Pull request merged**